From 31bed20e6b6988d951e08c4f2f019526412d9c3b Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Wed, 10 Mar 2021 16:15:17 -0700
Subject: [PATCH 24/24] blk-mq: Disable -Walign-mismatch in
 __blk_mq_complete_request

When compiling with clang-13:

block/blk-mq.c:630:39: warning: passing 8-byte aligned argument to
32-byte aligned parameter 2 of 'smp_call_function_single_async' may
result in an unaligned pointer access [-Walign-mismatch]
                smp_call_function_single_async(cpu, &rq->csd);
                                                    ^
1 warning generated.

This was purposefully done in commit 4ccafe032005 ("block: unalign
call_single_data in struct request") and there is no real bug here.
Disable the warning using the __diag_ignore infrastructure.

Link: https://github.com/ClangBuiltLinux/linux/issues/1328
Link: https://lore.kernel.org/r/20210310182307.zzcbi5w5jrmveld4@archlinux-ax161/
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 block/blk-mq.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 057a634396a9..d49f4fef2785 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -619,7 +619,10 @@ static void __blk_mq_complete_request(struct request *rq)
 		rq->csd.func = __blk_mq_complete_request_remote;
 		rq->csd.info = rq;
 		rq->csd.flags = 0;
+		__diag_push();
+		__diag_ignore(clang, 13, "-Walign-mismatch", "There is no issue with misalignment here");
 		smp_call_function_single_async(ctx->cpu, &rq->csd);
+		__diag_pop();
 	} else {
 		q->mq_ops->complete(rq);
 	}
-- 
2.31.0

